sycl/test-e2e/Matrix/joint_matrix_bfloat16_array_impl.hpp

```diff
@@ -86,35 +84,68 @@ void matrix_multiply(big_matrix<T1, M, N> &C, big_matrix<T2, M, K> &A,
   static constexpr size_t MATRIX_N = TN * 2;
   static constexpr size_t MATRIX_K = TK * 2;
 
-  bfloat16 A[MATRIX_M][MATRIX_K];
-  bfloat16 B[MATRIX_K / 2][MATRIX_N * 2];
+  T A[MATRIX_M][MATRIX_K];
+  T B[MATRIX_K / VNNI][MATRIX_N * VNNI];
 
-  float C[MATRIX_M][MATRIX_N];
-  float D[MATRIX_M][MATRIX_N];
+  TResult C[MATRIX_M][MATRIX_N];
+  TResult D[MATRIX_M][MATRIX_N];
 
-  matrix_fill(MATRIX_M, MATRIX_K, (bfloat16 *)A,
-              [](int i, int j) { return 1.0f * (i + j); });
-  matrix_fill(MATRIX_K / 2, MATRIX_N * 2, (bfloat16 *)B,
-              [](int i, int j) { return 2.0f * i + 3.0f * j; });
-  matrix_fill(MATRIX_M, MATRIX_N, (float *)C, 1.0f);
-  matrix_fill(MATRIX_M, MATRIX_N, (float *)D, 1.0f);
+  matrix_fill(MATRIX_M, MATRIX_K, (T *)A,
+              [](int i, int j) { return TResult(1) * (i + j); });
+  matrix_fill(MATRIX_K / VNNI, MATRIX_N * VNNI, (T *)B,
+              [](int i, int j) { return TResult(2) * i + TResult(3) * j; });
+  matrix_fill(MATRIX_M, MATRIX_N, (TResult *)C, TResult(1));
+  matrix_fill(MATRIX_M, MATRIX_N, (TResult *)D, TResult(1));
 
-  big_matrix<float, MATRIX_M, MATRIX_N> MC((float *)&C);
-  big_matrix<float, MATRIX_M, MATRIX_N> MD((float *)&D);
-  big_matrix<bfloat16, MATRIX_M, MATRIX_K> MA((bfloat16 *)&A);
-  big_matrix<bfloat16, MATRIX_K / 2, MATRIX_N * 2> MB((bfloat16 *)&B);
+  big_matrix<TResult, MATRIX_M, MATRIX_N> MC((TResult *)&C);
+  big_matrix<TResult, MATRIX_M, MATRIX_N> MD((TResult *)&D);
+  big_matrix<T, MATRIX_M, MATRIX_K> MA((T *)&A);
+  big_matrix<T, MATRIX_K / VNNI, MATRIX_N * VNNI> MB((T *)&B);
 
-  matrix_multiply(MC, MA, MB);
-  matrix_multiply_ref<bfloat16, bfloat16, float, 2>(
-      (bfloat16 *)A, (bfloat16 *)B, (float *)D, MATRIX_M, MATRIX_N,
-      MATRIX_K / 2);
+  matrix_multiply<TResult, T, MATRIX_M, MATRIX_N, MATRIX_K, TM, TN, TK, VNNI>(
+      MC, MA, MB);
+  matrix_multiply_ref<T, T, TResult, VNNI>((T *)A, (T *)B, (TResult *)D,
+                                           MATRIX_M, MATRIX_N, MATRIX_K / VNNI);
 
-  bool res = matrix_compare(MATRIX_M, MATRIX_N, (float *)C, (float *)D);
-  std::cout << (res ? "passed" : "failed") << std::endl;
-  return !res;
+  assert(matrix_compare(MATRIX_M, MATRIX_N, (TResult *)C, (TResult *)D));
 }
+
+int main() {
+  queue q;
+  std::vector<combination> combinations =
+      q.get_device()
+          .get_info<sycl::ext::oneapi::experimental::info::device::
+                        matrix_combinations>();
+
+  for (unsigned int i = 0; i < combinations.size(); i++) {
+    if (combinations[i].nsize == 0) { // Intel AMX
+      test<bfloat16, float, 2, /*TM*/ 16, /*TN*/ 16, /*TK*/ 32>();
+      break;
+    }
+
+    if (combinations[i].nsize == 16) { // architecture::intel_gpu_pvc
+      test<bfloat16, float, 2, /*TM*/ 8, /*TN*/ 16, /*TK*/ 16>();
+#if (!defined(SG_SZ) || SG_SZ != 32)
+      // These combination are not currently supported for subgroup size = 32 in
+      // IGC
+      test<bfloat16, float, 2, /*TM*/ 16, /*TN*/ 16, /*TK*/ 16>();
```

Comment:
I'm wondering, did you check that 16x16x16 doesn't work on PVC with SG=32?

We have at least one test, which works with this combination: https://github.com/intel/llvm/blob/sycl/sycl/test-e2e/Matrix/element_wise_all_ops_impl.hpp

So, if 16x16x16 works actually, then these tests also should be updated accordingly:
https://github.com/intel/llvm/blob/sycl/sycl/test-e2e/Matrix/element_wise_ops_impl.hpp
https://github.com/intel/llvm/blob/sycl/sycl/test-e2e/Matrix/joint_matrix_apply_bf16_impl.hpp
https://github.com/intel/llvm/blob/sycl/sycl/test-e2e/Matrix/joint_matrix_bfloat16_impl.hpp

Reply:
So...
* all combinations work for `bfloat16` with `SG_SZ=32` work for `joint_matrix_bfloat16_array_impl.hpp`
* I did test the other tests you mentioned, and they only pass with the default `SG_SZ`. The test produce two failure modes when the `#ifdef` is removed
``` 
error: undefined reference to `__builtin_spriv_OpJointMatrixStoreINTEL_PackedA_RowMajor_SG16_32x16_i16_16_global_pi64_v8i8'
in function: '__builtin_spriv_OpJointMatrixStoreINTEL_PackedA_RowMajor_SG16_32x16_i16_16_global_pi64_v8i8' called by kernels:
  - 'typeinfo name for add_func<sycl::_V1::ext::oneapi::bfloat16, 32ul, 16ul>'
  - 'typeinfo name for add<sycl::_V1::ext::oneapi::bfloat16, 32ul, 16ul>'
```
and 
```
terminate called after throwing an instance of 'sycl::_V1::compile_program_error'
  what():  The program was built for 1 devices
Build program log for 'Intel(R) Data Center GPU Max 1100':
 -11 (PI_ERROR_BUILD_PROGRAM_FAILURE)
```

Taking a look at the tests i failed to spot anything with the tests that might trigger this.

Reply:
The error that you show is for matrix A 32x16, it is not for A 16x16, did you test 16x16x16 specifically, or just all combinations?

Reply:
Verified in a call that all these tests fail with 16x16x16 combination on SG32. There was a typo in the cpp file that launched the test with SG16, instead of SG32